qiskit/circuit/library/standard_gates/x.py

```diff
@@ -924,21 +924,24 @@ def __new__(
         num_ctrl_qubits: Optional[int] = None,
         label: Optional[str] = None,
         ctrl_state: Optional[Union[str, int]] = None,
+        _name="mcx",
+        synthesis=None,
```

Comment:
What is the `_name` argument required for? 🙂 

Could we put the `synthesis` argument first? This seems to be the argument a user should use over the `_name` one. Also could you add type hints?

Reply:
@Cryoris Many thanks for your comments! 

The `_name` thing kind of existed before: it's the name given to the returned quantum circuit as per `qc = QuantumCircuit(q, name=self.name)`. Does it really matter what name it has? If not, sure, let's remove it. 

I need to significantly improve all comments, add type hints, make the code pass black/lint, and in particular resolve the cyclic import dependencies. The synthesis algorithms probably need to be aware of gates, yet somehow the gates also need to be aware of the synthesis algorithms (e.g., the `MCXGate.__init__`  sets the default synthesis algorithm to `MCXSynthesisGrayCode`).

Reply:
About type hints. If I understand correctly, you are suggesting to replace  `synthesis=None,` by `synthesis : Optional[MCXSynthesis] = None`. However, this leads to a chain of cyclic dependencies that I cannot easily resolve: the file `x.py` defining the MCX gate now requires the file `mcx_synthesis.py`, which requires `x.py`. This can be solved by moving the abstract MCX synthesis interface to a separate file. So, does it make sense to have two files? Does it makes sense to have a whole directory for MCX synthesis algorithms? 

I am also confused about the `_name` argument. Contrary to what I thought earlier, it's the name assigned to the gate, and currently it can be of "mcx", "mcx_gray", "mcx_recursive" or "mcx_vchain". Does it make sense to keep this difference, or would a single "mcx" be fine for all?

Reply:
> If I understand correctly, you are suggesting to replace synthesis=None, by synthesis : Optional[MCXSynthesis] = None

Yes, exactly. 🙂 

> Does it makes sense to have a whole directory for MCX synthesis algorithms?

Yes, I would say so, similar to `qiskit/synthesis/evolution`.

About the name, we have to be careful since the circuit simulators might use the name to determine the operation. I think we can leave the names as they are. Ideally, we would deprecate all gates except `MCXGate` since the others are available by choosing the appropriate synthesis, but we can tackle this in a follow-up.